cdap-app-templates/cdap-etl/hydrator-spark-core/src/main/java/co/cask/cdap/etl/spark/function/JoinMergeFunction.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.etl.spark.function;
+
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.cdap.etl.api.JoinElement;
+import co.cask.cdap.etl.api.Transformation;
+import co.cask.cdap.etl.api.batch.BatchJoiner;
+import co.cask.cdap.etl.api.batch.BatchJoinerRuntimeContext;
+import co.cask.cdap.etl.common.DefaultEmitter;
+import co.cask.cdap.etl.common.TrackedTransform;
+import org.apache.spark.api.java.function.FlatMapFunction;
+import scala.Tuple2;
+
+import java.util.List;
+
+/**
+ * Function that merges a join result using a BatchJoiner.
+ */
+public class JoinMergeFunction implements FlatMapFunction<Tuple2<Object, List<JoinElement<Object>>>, Object> {
+  private final PluginFunctionContext pluginFunctionContext;
+  private transient TrackedTransform<Tuple2<Object, List<JoinElement<Object>>>, Object> joinFunction;
+  private transient DefaultEmitter<Object> emitter;
+
+  public JoinMergeFunction(PluginFunctionContext pluginFunctionContext) {
+    this.pluginFunctionContext = pluginFunctionContext;
+  }
+
+  @Override
+  public Iterable<Object> call(Tuple2<Object, List<JoinElement<Object>>> input) throws Exception {
+    if (joinFunction == null) {
```

Comment:
yep. This function is run in a spark closure, which means Spark is responsible for constructing the object and populating the fields through deserialization. So all fields need to be serializable, or created through serializable fields.

I can add a comment to clarify.